File: apps/api/coinmark_api/services/market_supply.py
import asyncio
import time
from typing import Any

import httpx


_CACHE_TTL_SEC = 6 * 60 * 60
_cache: dict[str, dict[str, Any]] = {}
_cache_ts: float = 0.0
_lock = asyncio.Lock()


def _to_float(v) -> float | None:
    try:
        return float(v)
    except (TypeError, ValueError):
        return None


async def _fetch_apex_compliance() -> list[dict[str, Any]]:
    url = "https://www.binance.com/bapi/apex/v1/friendly/apex/marketing/complianceSymbolList"
    headers = {"User-Agent": "coinmark/1.0"}
    async with httpx.AsyncClient(timeout=8) as client:
        resp = await client.get(url, headers=headers)
        resp.raise_for_status()
        data = resp.json()
    if isinstance(data, dict):
        items = data.get("data")
        if isinstance(items, list):
            return items
        if isinstance(items, dict) and isinstance(items.get("data"), list):
            return items["data"]
    return []


async def _fetch_asset_products() -> list[dict[str, Any]]:
    url = "https://www.binance.com/exchange-api/v2/public/asset-service/product/get-products"
    params = {"includeEtf": "true"}
    headers = {"User-Agent": "coinmark/1.0"}
    async with httpx.AsyncClient(timeout=8) as client:
        resp = await client.get(url, params=params, headers=headers)
        resp.raise_for_status()
        data = resp.json()
    if isinstance(data, dict):
        items = data.get("data")
        if isinstance(items, list):
            return items
        if isinstance(items, dict) and isinstance(items.get("data"), list):
            return items["data"]
    return []


def _merge_snapshot(base: str, row: dict[str, Any], cur: dict[str, Any] | None) -> dict[str, Any]:
    if cur is None:
        return row
    for key in ("circulating_supply", "total_supply", "max_supply", "market_cap", "price"):
        if cur.get(key) is None and row.get(key) is not None:
            cur[key] = row[key]
    if cur.get("source") is None and row.get("source") is not None:
        cur["source"] = row["source"]
    return cur


async def _refresh_cache() -> None:
    global _cache_ts
    next_cache: dict[str, dict[str, Any]] = {}

    try:
        apex_rows = await _fetch_apex_compliance()
    except Exception:
        apex_rows = []

    for row in apex_rows:
        base = str(row.get("name") or "").upper()
        symbol_pair = str(row.get("symbol") or "").upper()
        if not base and symbol_pair.endswith("USDT"):
            base = symbol_pair[: -len("USDT")]
        if not base:
            continue
        entry = {
            "symbol_pair": symbol_pair or None,
            "circulating_supply": _to_float(row.get("circulatingSupply")),
            "total_supply": _to_float(row.get("totalSupply")),
            "max_supply": _to_float(row.get("maxSupply")),
            "market_cap": _to_float(row.get("marketCap")),
            "price": _to_float(row.get("price")),
            "source": "binance_apex",
        }
        cur = next_cache.get(base)
        if cur is None:
            next_cache[base] = entry
        else:
            cur_cap = cur.get("market_cap") or 0
            next_cap = entry.get("market_cap") or 0
            if next_cap > cur_cap:
                next_cache[base] = _merge_snapshot(base, entry, cur)
            else:
                next_cache[base] = _merge_snapshot(base, entry, cur)

    try:
        product_rows = await _fetch_asset_products()
    except Exception:
        product_rows = []

    for row in product_rows:
        base = str(row.get("b") or "").upper()
        quote = str(row.get("q") or "").upper()
        symbol_pair = str(row.get("s") or "").upper()
        if not base and symbol_pair.endswith("USDT"):
            base = symbol_pair[: -len("USDT")]
            quote = "USDT"
        if not base or (quote and quote != "USDT"):
            continue
        entry = {
            "symbol_pair": symbol_pair or None,
            "circulating_supply": _to_float(row.get("cs")),
            "market_cap": None,
            "price": _to_float(row.get("c")),
            "source": "binance_asset",
        }
        cur = next_cache.get(base)
        next_cache[base] = _merge_snapshot(base, entry, cur)

    _cache.clear()
    _cache.update(next_cache)
    _cache_ts = time.time()


async def get_supply_snapshot(symbol: str) -> dict[str, Any] | None:
    now = time.time()
    if now - _cache_ts > _CACHE_TTL_SEC or not _cache:
        async with _lock:
            if time.time() - _cache_ts > _CACHE_TTL_SEC or not _cache:
                await _refresh_cache()

    return _cache.get(symbol.upper())

File: apps/api-go/internal/service/market_supply.go
package service

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"strings"
	"sync"
	"time"
)

const supplyCacheTTL = 6 * time.Hour

type SupplySnapshot struct {
	SymbolPair        string   `json:"symbol_pair,omitempty"`
	CirculatingSupply *float64 `json:"circulating_supply,omitempty"`
	TotalSupply       *float64 `json:"total_supply,omitempty"`
	MaxSupply         *float64 `json:"max_supply,omitempty"`
	MarketCap         *float64 `json:"market_cap,omitempty"`
	Price             *float64 `json:"price,omitempty"`
	Source            string   `json:"source,omitempty"`
}

type SupplyCache struct {
	mu     sync.Mutex
	cache  map[string]*SupplySnapshot
	ts     time.Time
	client *http.Client
}

func NewSupplyCache() *SupplyCache {
	return &SupplyCache{
		cache:  make(map[string]*SupplySnapshot),
		client: &http.Client{Timeout: 8 * time.Second},
	}
}

func (sc *SupplyCache) Get(ctx context.Context, symbol string) (*SupplySnapshot, error) {
	sc.mu.Lock()
	if time.Since(sc.ts) > supplyCacheTTL || len(sc.cache) == 0 {
		sc.mu.Unlock()
		if err := sc.refresh(ctx); err != nil {
			return nil, err
		}
		sc.mu.Lock()
	}
	s := sc.cache[strings.ToUpper(symbol)]
	sc.mu.Unlock()
	return s, nil
}

func (sc *SupplyCache) refresh(ctx context.Context) error {
	sc.mu.Lock()
	if time.Since(sc.ts) <= supplyCacheTTL && len(sc.cache) > 0 {
		sc.mu.Unlock()
		return nil
	}
	sc.mu.Unlock()

	next := make(map[string]*SupplySnapshot)

	apex, _ := sc.fetchApexCompliance(ctx)
	for _, row := range apex {
		base := strings.ToUpper(str(row, "name"))
		pair := strings.ToUpper(str(row, "symbol"))
		if base == "" && strings.HasSuffix(pair, "USDT") {
			base = pair[:len(pair)-4]
		}
		if base == "" {
			continue
		}
		entry := &SupplySnapshot{
			SymbolPair:        pair,
			CirculatingSupply: optFloat(row, "circulatingSupply"),
			TotalSupply:       optFloat(row, "totalSupply"),
			MaxSupply:         optFloat(row, "maxSupply"),
			MarketCap:         optFloat(row, "marketCap"),
			Price:             optFloat(row, "price"),
			Source:            "binance_apex",
		}
		mergeSupply(next, base, entry)
	}

	products, _ := sc.fetchAssetProducts(ctx)
	for _, row := range products {
		base := strings.ToUpper(str(row, "b"))
		quote := strings.ToUpper(str(row, "q"))
		pair := strings.ToUpper(str(row, "s"))
		if base == "" && strings.HasSuffix(pair, "USDT") {
			base = pair[:len(pair)-4]
			quote = "USDT"
		}
		if base == "" || (quote != "" && quote != "USDT") {
			continue
		}
		entry := &SupplySnapshot{
			SymbolPair:        pair,
			CirculatingSupply: optFloat(row, "cs"),
			Price:             optFloat(row, "c"),
			Source:            "binance_asset",
		}
		mergeSupply(next, base, entry)
	}

	sc.mu.Lock()
	sc.cache = next
	sc.ts = time.Now()
	sc.mu.Unlock()
	return nil
}

func mergeSupply(m map[string]*SupplySnapshot, base string, entry *SupplySnapshot) {
	cur, ok := m[base]
	if !ok {
		m[base] = entry
		return
	}
	if cur.CirculatingSupply == nil && entry.CirculatingSupply != nil {
		cur.CirculatingSupply = entry.CirculatingSupply
	}
	if cur.TotalSupply == nil && entry.TotalSupply != nil {
		cur.TotalSupply = entry.TotalSupply
	}
	if cur.MaxSupply == nil && entry.MaxSupply != nil {
		cur.MaxSupply = entry.MaxSupply
	}
	if cur.MarketCap == nil && entry.MarketCap != nil {
		cur.MarketCap = entry.MarketCap
	}
	if cur.Price == nil && entry.Price != nil {
		cur.Price = entry.Price
	}
	if cur.Source == "" && entry.Source != "" {
		cur.Source = entry.Source
	}
}

func (sc *SupplyCache) fetchApexCompliance(ctx context.Context) ([]map[string]interface{}, error) {
	return sc.fetchJSON(ctx, "https://www.binance.com/bapi/apex/v1/friendly/apex/marketing/complianceSymbolList", nil)
}

func (sc *SupplyCache) fetchAssetProducts(ctx context.Context) ([]map[string]interface{}, error) {
	return sc.fetchJSON(ctx, "https://www.binance.com/exchange-api/v2/public/asset-service/product/get-products", map[string]string{"includeEtf": "true"})
}

func (sc *SupplyCache) fetchJSON(ctx context.Context, rawURL string, params map[string]string) ([]map[string]interface{}, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, rawURL, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("User-Agent", "coinmark/1.0")
	if len(params) > 0 {
		q := req.URL.Query()
		for k, v := range params {
			q.Set(k, v)
		}
		req.URL.RawQuery = q.Encode()
	}
	resp, err := sc.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("market_supply: %d %s", resp.StatusCode, string(body))
	}
	var raw map[string]interface{}
	if err := json.Unmarshal(body, &raw); err != nil {
		return nil, err
	}
	data := raw["data"]
	switch d := data.(type) {
	case []interface{}:
		return toMapSlice(d), nil
	case map[string]interface{}:
		if inner, ok := d["data"].([]interface{}); ok {
			return toMapSlice(inner), nil
		}
	}
	return nil, nil
}

func toMapSlice(items []interface{}) []map[string]interface{} {
	out := make([]map[string]interface{}, 0, len(items))
	for _, item := range items {
		if m, ok := item.(map[string]interface{}); ok {
			out = append(out, m)
		}
	}
	return out
}

func str(m map[string]interface{}, key string) string {
	v, _ := m[key].(string)
	return v
}

func optFloat(m map[string]interface{}, key string) *float64 {
	v, ok := m[key]
	if !ok || v == nil {
		return nil
	}
	switch t := v.(type) {
	case float64:
		return &t
	case string:
		f, err := strconv.ParseFloat(t, 64)
		if err != nil {
			return nil
		}
		return &f
	default:
		return nil
	}
}
